### ISSUE TITLE
SIGPIPE when OpenSSL is used with a socket

### ISSUE THREAD

**Issue by httpstorm** [openssl/openssl#16399]
Please provide a way for user apps to specify socket send and receive flags. On UNIX systems when the remote side terminates the socket, the user app will crash with `SIGPIPE`. On Linux, apps using sockets can specify the `MSG_NO_SIGNAL` flag in the `send` function, to avoid receiving `SIGPIPE`, however when OpenSSL is used, there is no way to any flags when sending and receiving from sockets. On macOS and BSD, the app can specify `SO_NOSIGPIPE` in `setsockopt` to overcome this issue.

OpenSSL 1.1.1k
Ubuntu 20.04

https://github.com/openssl/openssl/issues/8075#issuecomment-902145012

**Comment by pavel-odintsov** [comment]
We just caught this issue in our code. It's pretty serious issue for us as we cannot easily suppress SIGPIPE as we need to have visibility when it happens due to extensive use of popen() in our logic.

For plain TCP logic we use send() instead of write() and we set last flag argument of send() to MSG_NOSIGNAL and it works just fine.

Would be great to have some way to pass flags field to SSL_write().

Thank you!

**Comment by mattcaswell** [comment]
A workaround for this would be to create a custom BIO to do the socket writing instead (which means you can insert whatever flags you want in a `send` call). See the following man page for creating a custom BIO:

https://www.openssl.org/docs/man3.1/man3/BIO_meth_new.html

Using the custom BIO you set that on your `SSL` object using `SSL_set0_wbio()` (or similar function). Then, whenever OpenSSL wants to write to the socket, it will use your custom BIO code instead.

**Comment by Alcaro** [comment]
I had the same issue. My solution was to create a custom BIO method and tell OpenSSL to send/receive that way (BIO_meth_new, BIO_meth_set_write_ex, BIO_meth_set_read_ex, BIO_meth_set_ctrl (just return 0 for everything, except BIO_CTRL_FLUSH which must return 1), BIO_new, SSL_set_bio, and BIO_set_data/BIO_get_data to tell it which fd to use).

But that has the obvious side effects on kTLS support, and probably some other features.

**Comment by mattcaswell** [comment]
> @mattcaswell doesn't it make sense to allow overwriting readsocket/writesocket in standard BIOs?

Yes - that might make sense, with some ctrl enabling you to set what flags to use. But that really is a new feature - my suggestion was a workaround for what you can do right now. I've added the "help wanted" flag if someone wants to implement this.

**Comment by paulidale** [comment]
@mattcaswell wouldn't it be possible to use `BIO_new_socket()` and `SSL_set0_[rw]rbio()` here?  Set the socket option before calling the former.

**Comment by mattcaswell** [comment]
> @mattcaswell wouldn't it be possible to use BIO_new_socket() and SSL_set0_[rw]rbio() here? Set the socket option before calling the former.

Unfortunately, `MSG_NOSIGNAL` isn't a socket option. It's a flag on the `send` call - so it has to be provided every time you try to write something.

**Comment by paulidale** [comment]
Huh?  The original issue includes this:

> On macOS and BSD, the app can specify SO_NOSIGPIPE in setsockopt to overcome this issue.

**Comment by Alcaro** [comment]
MSG_NOSIGNAL is not SO_NOSIGPIPE. There is no equivalent of SO_NOSIGPIPE on Linux.

**Comment by ingo-h** [comment]
I'm also fighting with the SIGPIPE problem for a library that must not alter the users process.

SO_NOSIGPIPE does not help on macOS and BSD to workaround the SIGPIPE problem with OpenSSL. It has no effect on system call `write()` that is used by OpenSSL. You can find a note at [SIGPIPE and how to ignore it](https://www.doof.me.uk/2020/09/23/sigpipe-and-how-to-ignore-it/)

> This is a bit niche as it only exists on FreeBSD and OS X. Use setsockopt(2) to set this option on a socket and all subsequent send(2) calls will behave as if MSG_NOSIGNAL was set. [..] This seems to be of limited utility as calling write(2) on the socket will still generate SIGPIPE.

With `man send` you will find:

> The only difference between send() and write(2) is the presence of flags. With a zero flags argument, send() is equivalent to write(2).

So my suggestion is to use `send()` instead of `write()` in OpenSSL and and make setting the flags available to the user. This also help on Linux platforms with setting MSG_NOSIGNAL that isn't possible with `write()`.

**Comment by ingo-h** [comment]
@pavel-odintsov  I just see that you are already partially using send() https://github.com/openssl/openssl/issues/16399#issuecomment-1785649956. This could easily(?) be expanded?

**Comment by ingo-h** [comment]
For a workaround now I have made a solution with general suppressing the SIGPIPE signal temporary on the current thread. For those who are interested may have a look at [C++ & OpenSSL: SIGPIPE when writing in closed pipe](https://stackoverflow.com/a/77688810/5014688).

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by pavel-odintsov** [comment]
Hello!

I can confirm that this feature is very important for my company

Sincerely yours, Pavel Odintsov

**Comment by ingo-h** [comment]
As already noted in my comments earlier I use a complicated workaround. It would be very helpful, much easier, and less error prone to have this feature.

As shown in my https://github.com/openssl/openssl/issues/16399#issuecomment-1858939775 I suggest to use send() instead of write() in OpenSSL and make setting the flags available to the user. This is a simple replace of system function calls that do the same with the possibility to set flags. The only additional function to implement is to manage the flags. Isn't it?

**Comment by rlebeau** [comment]
This only addresses the issue for socket sends, but what about for socket reads, too? Encountering an issue where a server is throwing a SIGPIPE on a socket read after a client resets the connection.

**Comment by Sir-Walrus** [comment]
The read() syscall does not yield SIGPIPE ([source](https://godbolt.org/z/4TYT3MYxs)).

Maybe you're calling a read function in OpenSSL, but OpenSSL does some writes too, due to heartbeat or renegotiation or something? We're gonna need more information than that, for example a backtrace or strace log.

**Comment by rlebeau** [comment]
> Maybe you're calling a read function in OpenSSL, but OpenSSL does some writes too, due to heartbeat or renegotiation or something?

That's a good point.
